### DISCUSSION TITLE
Structured movements

### DISCUSSION THREAD

**Discussion by losnappas** [helix-editor/helix#7073] [Ideas]
Bringing up this idea from emacs' [symex](https://github.com/drym-org/symex.el/tree/2.0-integration/) plugin. Essentially, any language with tree sitter can have navigation based on structures. E.g. `k` might select the next sibling in the structure.

There's already a couple of tree sitter movements `alt-i/o` in helix, I'm wondering why there isn't a good bunch more. I didn't bother checking if I could already make a keymap like this, but if anybody has then please share!

With tree sitter, there's also the opportunity to do simple refactors like "raise form" to lift something out of an `if` statement.

These will probably work good as a plugin, once the time comes. I hope plugins will have access to tree sitter structures and LSP.

**Comment by hamrik** [comment]
Use can use [unimparied mode](https://docs.helix-editor.com/keymap.html#unimpaired) to jump between TS nodes and `Alt-i/o` to expand/shrink their selection.

**Reply by losnappas** [reply]
That's a start. What I'd also imagine useful is simple refactors like `slurp` (familiar from working with lisps) working in any tree-sitter supported lang.

E.g.

```python
if something: # if this block is selected and you "slurp", then print(2) moves into the if statement
  print(1)
print(2)

# post-slurp:
if something:
  print(1)
  print(2)
```

I think this should be possible.

**Comment by valpackett** [comment]
[Ki editor](https://ki-editor.github.io/ki-editor/docs/pitch/) is out there now prioritizing tree-sitter based structured **editing**. Deleting/duplicating/reordering syntax nodes is VERY cool, I really want Helix to be able to do that.
